src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java

```diff
@@ -304,4 +320,97 @@ public void onClosed(Channel channel, IOException cause) {
       logException(computer, listener, "Error: ", e);
     }
   }
+
+  protected Connection connectToSsh(ComputeEngineComputer computer, TaskListener listener)
+      throws Exception {
+    ComputeEngineInstance node = computer.getNode();
+    if (node == null) {
+      throw new IllegalArgumentException("A ComputeEngineComputer with no node was provided");
+    }
+
+    final long timeout = node.getLaunchTimeoutMillis();
+    final long startTime = System.currentTimeMillis();
+    while (true) {
+      try {
+        long waitTime = System.currentTimeMillis() - startTime;
+        if (timeout > 0 && waitTime > timeout) {
+          // TODO(google-compute-engine-plugin/issues/135): better exception
+          throw new Exception(
+              "Timed out after "
+                  + (waitTime / 1000)
+                  + " seconds of waiting for ssh to become available. (maximum timeout configured is "
+                  + (timeout / 1000)
+                  + ")");
+        }
+        Instance instance = computer.refreshInstance();
+
+        String host = "";
+
+        // TODO(google-compute-engine-plugin/issues/136): handle multiple NICs
+        NetworkInterface nic = instance.getNetworkInterfaces().get(0);
+
+        if (this.useInternalAddress) {
+          host = nic.getNetworkIP();
+        } else {
+          // Look for a public IP address
+          if (nic.getAccessConfigs() != null) {
+            for (AccessConfig ac : nic.getAccessConfigs()) {
+              if (ac.getType().equals(InstanceConfiguration.NAT_TYPE)) {
+                host = ac.getNatIP();
+              }
+            }
+          }
+          // No public address found. Fall back to internal address
+          if (host.isEmpty()) {
+            host = nic.getNetworkIP();
+          }
+        }
+
+        int port = SSH_PORT;
+        logInfo(
+            computer,
+            listener,
+            "Connecting to " + host + " on port " + port + ", with timeout " + SSH_TIMEOUT + ".");
+        Connection conn = new Connection(host, port);
+        ProxyConfiguration proxyConfig = Jenkins.get().proxy;
+        Proxy proxy = proxyConfig == null ? Proxy.NO_PROXY : proxyConfig.createProxy(host);
+        if (!node.isIgnoreProxy()
+            && !proxy.equals(Proxy.NO_PROXY)
+            && proxy.address() instanceof InetSocketAddress) {
+          InetSocketAddress address = (InetSocketAddress) proxy.address();
+          HTTPProxyData proxyData = null;
+          if (proxyConfig.getUserName() != null) {
```

Comment:
Should probably also check to ensure that the password exists since it's used below as well. What's the expected behavior when username == "", or when username ="some value" && password == null?

Reply:
From what I've found, username and password must both be supplied for authentication. Also, if a username or password are not provided when submitting the form (those text boxes are left empty) it is intended that it will be null. I also verified that the empty string for username isn't possible here on my jenkins master.